File: testClients/kellogg/AI.py
#-*-python-*-
from BaseAI import BaseAI
from GameObject import *
import random
import math
import heapq
		
class AI(BaseAI):
  """The class implementing gameplay logic."""
  @staticmethod
  def username():
    return "Kellogg"

  @staticmethod
  def password():
    return "password"

  ##This function is called once, before your first turn
  def init(self):
    pass
 
  ##This function is called once, after your last turn
  def end(self):
    pass

  def moveTo(self,creature,target):
     path = self.pathFind(creature.x,creature.y,target.x,target.y)
     if path != None:
       while creature.movementLeft>0 and len(path)>0:
         next = path.pop()
         if next!=(creature.x,creature.y):
           creature.move(next[0],next[1])
      
  def distance(self,sourceX,sourceY,destX,destY):
    return int(math.sqrt((sourceX-destX)**2+(sourceY-destY)**2))
    
  def maxStat(self,creature):
    return max([creature.herbivorism,creature.carnivorism,creature.speed,creature.energy,creature.defense])

#TODO FIX    
  def findMate(self,creature):
      pass
        
  def getObject(self,x,y):
    return [lifeform for lifeform in self.creatures+self.plants if lifeform.x == x and lifeform.y == y]
    
  def findNearest(self,source,list):
    d = {self.distance(source.x,source.y,lifeform.x,lifeform.y):lifeform for lifeform in list if lifeform.id != source.id}
    return d[min(d)]    
  
  
  def adjacent(self,x,y,points):
    walls = [(lifeform.x,lifeform.y) for lifeform in self.plants+self.creatures if (lifeform.x,lifeform.y) not in points or isinstance(lifeform,Plant) and lifeform.size==0]
    adj = []
    if x+1<self.mapWidth and (x+1,y) not in walls:
        adj.append((x+1,y))
    if y-1>=0 and (x,y-1) not in walls:
        adj.append((x,y-1))
    if x-1>=0 and (x-1,y) not in walls:
        adj.append((x-1,y))
    if y+1<self.mapHeight and (x,y+1) not in walls:
        adj.append((x,y+1))
    return adj
  
  def pathFind(self,startX,startY,goalX,goalY):
    closedSet = set();closedTup=set()
    open = [(self.distance(startX,startY,goalX,goalY),(startX,startY),(startX,startY),0)];openTup=[(startX,startY)]
    path = []#;ii=0
    while len(open)>0:
        open.sort()
        current = open[0]
        if current[1] == (goalX,goalY):
            node = current
            path = []
            while node[2]!=(startX,startY):
#              print "ii",ii
#              ii+=1
              for closed in closedSet:
#                print len(closedSet)
                if node[2] == closed[1]:
                  path.append(node[2])
                  node = closed
            return path
        closedSet.add(current);closedTup.add(current[1])
        open.remove(current); openTup.remove(current[1])
        for neighbor in self.adjacent(current[1][0],current[1][1],[(startX,startY),(goalX,goalY)]):
          if neighbor in closedTup:
          #if neighbor in [b[1] for b in closedSet]:
           continue
#          print current[3],neighbor[0],neighbor[1],current[1][0],current[1][1]
          g = current[3]+self.distance(neighbor[0],neighbor[1],current[1][0],current[1][1])
          if neighbor == (goalX,goalY) or self.distance(neighbor[0],neighbor[1],startX,startY)<=g+1 and neighbor not in openTup:
#          if self.distance(neighbor[0],neighbor[1],startX,startY)<=g+1 and neighbor not in [b[1] for b in open]:
            neighborTup = (g+self.distance(neighbor[0],neighbor[1],goalX,goalY),(neighbor[0],neighbor[1]),(current[1]),g)
            open.append(neighborTup);openTup.append(neighbor)
    return None
    
  ##This function is called each time it is your turn
  ##Return true to end your turn, return false to ask the server for updated information
  def run(self):   
    adjacent = [[1,0],[-1,0],[0,1],[0,-1]]    
    herbivores = [creature for creature in self.creatures if creature.owner == self.playerID and creature.herbivorism == self.maxStat(creature)]  
    carnivores = [creature for creature in self.creatures if creature.owner == self.playerID and creature.carnivorism == self.maxStat(creature)] 
    
    for creature in self.creatures:
     if creature.owner == self.playerID:             
       randPlant = self.plants[random.randrange(-1,len(self.plants))];room = True
       for adj in adjacent:
         if len(self.getObject(randPlant.x+adj[0],randPlant.y+adj[1]))>0:
           room = False
       if room:
         self.moveTo(creature,self.plants[self.playerID])
      # randx=random.randrange(-1,2); 
      # randy = abs(randx)^1*((-1)**random.randrange(1,100)%2)
      # x=0;y=0
      # if self.getObject(creature.x+randx,creature.y+randy) is None and (0<creature.x+randx<self.mapWidth) and (0<creature.y+randy<self.mapHeight):
        # creature.move(creature.x+randx,creature.y+randy)
      # for location in adjacent:
        # thingList=self.getObject(creature.x+location[0],creature.y+location[1]); thing = None
        # if len(thingList)>0:
          # thing = thingList[0]
        # if isinstance(thing,Plant) and thing.size>0 and creature.canEat:
         # creature.eat(thing.x,thing.y)
        # if isinstance(thing,Creature):
          # if thing.owner==self.playerID and thing.currentHealth > self.healthPerBreed and creature.currentHealth > self.healthPerBreed:
            # creature.breed(thing)
          # elif creature.canEat:
            # creature.eat(thing.x,thing.y)
    return 1

  def __init__(self, conn):
    BaseAI.__init__(self, conn)

File: testClients/Tyler_cpp/AI.cpp
#include "AI.h"
#include "util.h"
#include <cstdlib>
#include <cmath>

AI::AI(Connection* conn) : BaseAI(conn) {}

const char* AI::username()
{
  return "darklordtyler";
}

const char* AI::password()
{
  return "password";
}
bool AI::isValid(int x, int y)
{
   bool valid = true;
   if(x >= mapWidth() || x < 0 || y >= mapHeight() || y < 0)
   {
      valid = false;
   }
   for(int i = 0; i < mappables.size(); i++)
   {
      if(mappables[i].x() == x && mappables[i].y() == y)
      {
         
         valid = false;
      }
   }
   return valid;
}
/*
float distance(float x1, float x2, float y1, float y2)
{
   return sqrt(pow(x1 - x2, 2) + pow(y1 - y2, 2));
}
*/
Creature* AI::findAdjacentEnemy(Creature* & c)
{
   int creaturex = c -> x();
   int creaturey = c -> y();
   Creature* tmp = NULL;
   
   for(int i = 0; i < creatures.size(); i++)
   {
      if(creatures[i].owner() != playerID())
      {
         if(creatures[i].x() == creaturex + 1 && creatures[i].y() == creaturey) // right
         {
            tmp = &creatures[i];
         }
         if(creatures[i].x() == creaturex - 1 && creatures[i].y() == creaturey) // left
         {
            tmp = &creatures[i];
         }
         if(creatures[i].x() == creaturex && creatures[i].y() == creaturey + 1) // down
         {
            tmp = &creatures[i];
         }
         if(creatures[i].x() == creaturex && creatures[i].y() == creaturey - 1) // up
         {
            tmp = &creatures[i];
         }
      }
   }
   
   return tmp;
}

Creature* AI::findAdjacentFriend(Creature* & c)
{
   int creaturex = c -> x();
   int creaturey = c -> y();
   Creature* tmp = NULL;
   
   for(int i = 0; i < creatures.size(); i++)
   {
      if(creatures[i].owner() == playerID())
      {
         if(creatures[i].x() == creaturex + 1 && creatures[i].y() == creaturey) // right
         {
            tmp = &creatures[i];
         }
         if(creatures[i].x() == creaturex - 1 && creatures[i].y() == creaturey) // left
         {
            tmp = &creatures[i];
         }
         if(creatures[i].x() == creaturex && creatures[i].y() == creaturey + 1) // down
         {
            tmp = &creatures[i];
         }
         if(creatures[i].x() == creaturex && creatures[i].y() == creaturey - 1) // up
         {
            tmp = &creatures[i];
         }
      }
   }
   
   return tmp;
}

Plant* AI::findAdjacentPlant(Creature* & c)
{
   int creaturex = c -> x();
   int creaturey = c -> y();
   Plant* tmp = NULL;
   
   for(int i = 0; i < plants.size(); i++)
   {
      if(plants[i].x() == creaturex + 1 && plants[i].y() == creaturey) // right
      {
         tmp = &plants[i];
      }
      if(plants[i].x() == creaturex - 1 && plants[i].y() == creaturey) // left
      {
         tmp = &plants[i];
      }
      if(plants[i].x() == creaturex && plants[i].y() == creaturey + 1) // down
      {
         tmp = &plants[i];
      }
      if(plants[i].x() == creaturex && plants[i].y() == creaturey - 1) // up
      {
         tmp = &plants[i];
      }
   }
   
   return tmp;
}

bool AI::goLeft(Creature* & c)
{
   if(isValid(c -> x() - 1, c -> y()))
   {
      c -> move(c -> x() - 1, c -> y());
      return true;
   }
   return false;
}

bool AI::goRight(Creature* & c)
{
   if(isValid(c -> x() + 1, c -> y()))
   {
      c -> move(c -> x() + 1, c -> y());
      return true;
   }
   return false;
}

bool AI::goUp(Creature* & c)
{
   if(isValid(c -> x(), c -> y() - 1))
   {
      c -> move(c -> x(), c -> y() - 1);
      return true;
   }
   return false;
}

bool AI::goDown(Creature* & c)
{
   if(isValid(c -> x(), c -> y() + 1))
   {
      c -> move(c -> x(), c -> y() + 1);
      return true;
   }
   return false;
}

void AI::stupidMove(Creature* & c)
{
   int moves = 0;
   int max = (c -> movementLeft() / 2) + 1;
   int centerx = mapWidth() / 2;
   int centery = mapHeight() / 2;
   bool right = true;
   bool up = false;
   bool left = false;
   bool down = false;
   int badrun = 2;
   int trial = 0;
   
   while(moves < max)
   {
      int x = c -> x();
      int y = c -> y();
      
      if((centery - y > centerx - x && centery - y > 0 && trial == 0) || (badrun == 4 && trial != 0))
      {
         if(goDown(c))
         {
            moves++;
            trial = 0;
            badrun = 2;
         }
         else
         {
            badrun = 1;
            trial++;
         }
      }
      else if((centerx - x > centery - y && centerx - x > 0 && trial == 0) || (badrun == 1 && trial != 0))
      {
         if(goRight(c))
         {
            moves++;
            trial = 0;
            badrun = 2;
         }
         else
         {
            badrun = 2;
            trial++;
         }
      }
      else if((centerx - x > centery - y && centery - y < 0 && trial == 0) || (badrun == 2 && trial != 0))
      {
         if(goUp(c))
         {
            moves++;
            trial = 0;
            badrun = 2;
         }
         else
         {
            badrun = 3;
            trial++;
         }
      }
      else if((((centery - y > centerx - x && centerx - x < 0) || centerx - x == centery - y) && trial == 0) || (badrun == 3 && trial != 0))
      {
         if(goLeft(c))
         {
            moves++;
            trial = 0;
            badrun = 2;
         }
         else
         {
            badrun = 4;
            trial++;
         }
      }
      else
      {
         return;
      }
   }
}


//This function is run once, before your first turn.
void AI::init(){}


//This function is called each time it is your turn.
//Return true to end your turn, return false to ask the server for updated information.
bool AI::run()
{
  Creature* myCreatures[100];
  Creature* enemies[100];
  Creature* bestCreatures[5];
  int numMyCreatures = 0;
  int numEnemies = 0;
  int position1 = 0;
  int position2 = 0;
  // dudes with max stats
  int maxStrength = 0;
  int maxHerbivore = 0;
  int maxCarnivore = 0;
  int maxSpeedster = 0;
  int maxShield = 0;
  int bestlocx = mapWidth() / 2;
  int bestlocy = mapHeight() / 2;
  int plantx = -1;
  int planty = -1;
  float dist = mapWidth();
   
  for(int i = 0; i < creatures.size(); i++)
  {
    if(creatures[i].owner() == playerID())
    {
      myCreatures[position1] = &creatures[i];
      numMyCreatures++;
      position1++;
      int health = creatures[i].maxHealth();
      int herb = creatures[i].herbivorism();
      int carn = creatures[i].carnivorism();
      int speed = creatures[i].speed();
      int def = creatures[i].defense();
      if(health > maxStrength)
      {
         bestCreatures[0] = &creatures[i];
      }
      if(herb > maxHerbivore)
      {
         bestCreatures[1] = &creatures[i];
      }
      if(carn > maxCarnivore)
      {
         bestCreatures[2] = &creatures[i];
      }
      if(speed > maxSpeedster)
      {
         bestCreatures[3] = &creatures[i];
      }
      if(def > maxShield)
      {
         bestCreatures[4] = &creatures[i];
      }
    }
    else
    {
      enemies[position2] = &creatures[i];
      numEnemies++;
      position2++;
    }
  }
  /*
  for(int i = 0; i < plants.size(); i++)
  {
    float plantdist = distance(plants[i].x(), bestlocx, plants[i].y(), bestlocy);
    if(plantdist < dist)
    {
      dist = plantdist;
      plantx = plants[i].x();
      planty = plants[i].y();
    }
  }
  
  float munchies = distance(bestCreatures[1] -> x(), plantx, bestCreatures[1] -> y(), planty);
  if(munchies == 1)
  {
    bestCreatures[1] -> eat(plantx, planty);
  }
  else
  {
    int movesLeft = bestCreatures[1] -> movementLeft() / 3;
    while(movesLeft > 0 && munchies > 1)
    {
      pathfind(bestCreatures[1], bestCreatures[1] -> x(), bestCreatures[1] -> y(), plantx, planty, movesLeft);
    }
  }
  */
  
  for(int i = 0; i < numMyCreatures; i++)
  {
    int movesLeft = myCreatures[i] -> movementLeft();
    int xloc = myCreatures[i] -> x();
    int yloc = myCreatures[i] -> y();
    int health = myCreatures[i] -> maxHealth();
    int herb = myCreatures[i] -> herbivorism();
    int carn = myCreatures[i] -> carnivorism();
    int speed = myCreatures[i] -> speed();
    int def = myCreatures[i] -> defense();
    Creature* food = NULL;
    Creature* mate = NULL;
    Plant* shrub = NULL;
    
    if(myCreatures[i] -> canEat())
    {
      food = findAdjacentEnemy(myCreatures[i]);
      shrub = findAdjacentPlant(myCreatures[i]);
      if(food != NULL)
      {
         myCreatures[i] -> eat(food -> x(), food -> y());
      }
      else if(shrub != NULL)
      {
         myCreatures[i] -> eat(shrub -> x(), shrub -> y());
      }
    }
    
    stupidMove(myCreatures[i]);
    
    if(myCreatures[i] -> canBreed())
    {
      mate = findAdjacentFriend(myCreatures[i]);
      if(mate != NULL)
      {
         myCreatures[i] -> breed(*mate);
         stupidMove(mate);
      }
    }
    stupidMove(myCreatures[i]);
    
   
  }
  
  return true;
}

//This function is run once, after your last turn.
void AI::end(){}
